### PR TITLE
Add source_in_ring args for rings API

### PR DESCRIPTION
**Pull request by zhoney**
fixed: #523

Change-Id: If8c5cc4d8678f847ce4a5f0388e56aae465d3ea6

### CODE REVIEW

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/RaysAPI.java [hunk 1]
```diff
@@ -72,7 +72,7 @@ public String get(@Context GraphManager manager,
                       @DefaultValue(DEFAULT_PATHS_LIMIT) long limit) {
         LOG.debug("Graph [{}] get rays paths from '{}' with " +
                   "direction '{}', edge label '{}', max depth '{}', " +
-                  "max degree '{}' and limit '{}'",
+                  "max degree '{}', capacity '{}' and limit '{}'",
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/RaysAPI.java]: also add capacity to params list

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/RingsAPI.java [hunk 2]
```diff
@@ -72,7 +74,8 @@ public String get(@Context GraphManager manager,
                       @DefaultValue(DEFAULT_PATHS_LIMIT) long limit) {
         LOG.debug("Graph [{}] get rings paths reachable from '{}' with " +
                   "direction '{}', edge label '{}', max depth '{}', " +
-                  "max degree '{}' and limit '{}'",
+                  "source in ring '{}', max degree '{}', capacity '{}' " +
+                  "and limit '{}'",
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/RingsAPI.java]: ditto

File: hugegraph-api/src/main/java/com/baidu/hugegraph/version/ApiVersion.java [hunk 3]
```diff
@@ -84,10 +84,11 @@
      * [0.36] Issue-360: Support paging for scan api
      * [0.37] Issue-391: Add skip_super_node for shortest path
      * [0.38] Issue-274: Add personalrank and neighborrank RESTful API
+     * [0.39] Issue-523: Add sourceInRing args for rings RESTful API
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/version/ApiVersion.java]: Add source_in_ring arg for rings RESTful API

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java [hunk 4]
```diff
@@ -379,7 +380,8 @@ public boolean equals(Object object) {
                 return false;
             }
             Node other = (Node) object;
-            return this.id.equals(other.id);
+            return this.id.equals(other.id) &&
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java]: Objects.equals(this.id, other.id)

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java [hunk 5]
```diff
@@ -128,33 +154,78 @@ public Traverser(Id sourceV, Id label, long degree,
...
                             continue;
                         }
-                        // Rings found and expect rings
-                        if (this.rings) {
-                            assert n.contains(target);
-                            List<Id> prePath = n.path();
-                            prePath.add(target);
-                            paths.add(new Path(null, prePath));
+
+                        // Rays found if fake ring like:
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java]: if it's fake ring like

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java [hunk 6]
```diff
@@ -64,41 +68,61 @@ public SubGraphTraverser(HugeGraph graph) {
...
-            if (reachDepth || traverser.reachLimit() ||
+            if (--depth <= 0 || traverser.reachLimit() ||
                 traverser.finished()) {
                 break;
             }
         }
         return paths;
     }
 
+    private static boolean multiEdges(List<Edge> edges, Id target) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java]: containsMultiTimes

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java [hunk 7]
```diff
@@ -128,33 +154,78 @@ public Traverser(Id sourceV, Id label, long degree,
...
-                        if (this.rings) {
-                            assert n.contains(target);
-                            List<Id> prePath = n.path();
-                            prePath.add(target);
-                            paths.add(new Path(null, prePath));
+
+                        // Rays found if fake ring like:
+                        // path is pattern: A->B<-A && A is only neighbor of B
+                        if (!this.rings && target.equals(node.parent().id()) &&
+                            neighborCount == 1 && !edges.hasNext() &&
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java]: too many conditions

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java [hunk 8]
```diff
@@ -64,41 +68,61 @@ public SubGraphTraverser(HugeGraph graph) {
...
             }
         }
         return paths;
     }
 
+    private static boolean containsMultiEdges(List<Edge> edges, Id target) {
+        int count = 0;
+        for (Edge edge : edges) {
+            if (((HugeEdge) edge).id().otherVertexId().equals(target)) {
+                if (++count >= 2) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java]: use generalized `++count > 1`

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java [hunk 9]
```diff
@@ -128,33 +154,81 @@ public Traverser(Id sourceV, Id label, long degree,
...
+                            boolean ringsFound = false;
+                            // 1. sourceInRing is false, or
+                            // 2. sourceInRing is true and target == source
+                            if (!sourceInRing || target.equals(this.source)) {
+                                if (!target.equals(node.parent().id())) {
+                                    ringsFound = true;
+                                } else if (direction != Directions.BOTH) {
+                                    ringsFound = true;
+                                } else if (containsMultiEdges(edgeList,
+                                                              target)) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java]: align

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java [hunk 10]
```diff
@@ -64,41 +68,61 @@ public SubGraphTraverser(HugeGraph graph) {
...
+    private static boolean hasMultiEdges(List<Edge> edges, Id target) {
+        int count = 0;
+        for (Edge edge : edges) {
+            if (((HugeEdge) edge).id().otherVertexId().equals(target)) {
+                if (++count > 1) {
+                    return true;
+                }
+            }
+        }
+        assert count == 1;
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java]: seems count maybe = 0

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java [hunk 11]
```diff
@@ -112,7 +136,9 @@ public Traverser(Id sourceV, Id label, long degree,
             // Traversal vertices of previous level
             for (Map.Entry<Id, List<Node>> entry : this.sources.entrySet()) {
                 Id vid = entry.getKey();
-                edges = edgesOfVertex(vid, direction, this.label, this.degree);
+                List<Edge> edgeList = IteratorUtils.list(edgesOfVertex(
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java]: add some comment on `iterator -> list -> iterator`

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java [hunk 12]
```diff
@@ -128,33 +154,80 @@ public Traverser(Id sourceV, Id label, long degree,
...
                     }
                 }
+
+                int neighborCount = 0;
+                Set<Id> currentNeighbors = new HashSet<>();
                 while (edges.hasNext()) {
+                    neighborCount++;
                     HugeEdge edge = (HugeEdge) edges.next();
                     Id target = edge.id().otherVertexId();
+                    // Avoid dedup path
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java]: Avoid duplicate path

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java (hunk 10 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java]: there already exists one edge connected to 'target', so count must be >= 1, false means count == 1

**Pull request merged**